### PR TITLE
fix sort and topk with discontiguous out

### LINKED ISSUE #62645: Wrong results for `sort` kernel with noncontiguous outputs on CUDA.
## 🐛 Bug

`sort` CUDA kernel gives wrong results when calling its out variant for non-contiguous "values" output tensor.

## To Reproduce

```python
>>> a = torch.arange(12, device='cuda')
>>> indices = torch.zeros_like(a, dtype=torch.long)
>>> values = torch.zeros_like(a)
>>> # a, indices: get the first 6 elements
>>> # values: get a only the even positions
>>> torch.sort(a[:6], 0, out=(values[::2], indices[:6])) 
torch.return_types.sort_out(
values=tensor([0, 0, 1, 3, 4, 5], device='cuda:0'),
indices=tensor([5, 0, 1, 3, 2, 4], device='cuda:0'))
```

## Expected behavior

```python
>>> torch.sort(a[:6], 0, out=(values[::2], indices[:6])) 
torch.return_types.sort_out(
values=tensor([0, 1, 2, 3, 4, 5], device='cuda:0'),
indices=tensor([0, 1, 2, 3, 4, 5], device='cuda:0'))
```

## Environment

PyTorch version: 1.9.0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.9.5 (default, Jun  4 2021, 12:28:51)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.15.0-147-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: TITAN RTX
GPU 1: TITAN RTX

Nvidia driver version: 460.27.04
cuDNN version: /usr/local/cuda-10.2.89/targets/x86_64-linux/lib/libcudnn.so.7
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.2
[pip3] torch==1.9.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] mkl                       2021.2.0           h06a4308_296
[conda] mkl-service               2.3.0            py39h27cfd23_1
[conda] mkl_fft                   1.3.0            py39h42c9631_2
[conda] mkl_random                1.2.1            py39ha9443f7_2
[conda] numpy                     1.20.2           py39h2d18471_0
[conda] numpy-base                1.20.2           py39hfae3a4d_0
[conda] pytorch                   1.9.0           py3.9_cuda10.2_cudnn7.6.5_0    pytorch

## Additional context

I think the problem may be somewhere in the `sortKeyValueInplace` function (an inplace implementation), given that there are only a few elements to sort (`should_use_small_sort` is true) and that `sort` is always unstable.

https://github.com/pytorch/pytorch/blob/fd8004b42e2a2348ec8837e3fb524b960c1b4cdb/aten/src/ATen/native/cuda/Sort.cu#L264-L266

cc @ngimel @heitorschueroff

### PR DESCRIPTION
**Pull request by ngimel**
Fixes #62645 and #62940. The root cause of those bugs is in the bad interaction between `collapseDims` and setting the size of sorting/topK dimension to 1. If all other dimensions happen to be 1, `collapseDims` thinks that that `1` dimension is collapsible (even though it was specifically marked to be preserved) and loses its stride information. If dimension was really of size 1, the stride information would be unimportant, but since in reality that dimension is not 1 and was set to 1 for convenience, the loss of stride information results in incorrect outputs.

### CODE REVIEW

**Review by heitorschueroff — APPROVED** [review]
LGTM!
